src/traversals/dfs.jl

```diff
@@ -7,10 +7,28 @@
 Return `true` if graph `g` contains a cycle.
 
 ### Implementation Notes
-Uses DFS.
+The algorithm uses a DFS.
 """
 function is_cyclic end
-@traitfn is_cyclic(g::::(!IsDirected)) = ne(g) > 0
+@enum Vertex_state unvisited visited
+@traitfn function is_cyclic(g::AG::(!IsDirected)) where {T, AG<:AbstractGraph{T}}
+    visited = falses(nv(g)) 
+    for v in vertices(g)
+        visited[v] && continue
+        visited[v] = true
+        S = [(v,v)] 
+        while !isempty(S)
+            parent, w = pop!(S)
+            for u in neighbors(g, w)
+                u == parent && continue
+                visited[u] && return true
+                visited[u] = true
+                push!(S, (w, u))
+            end
+        end
+    end
+    return false
+end
```

Comment:
This looks correct to me, but there is a small edge that we also have to consider: If an undirected graph has a self-loop (i.e. a vertex that is connected to itself), then this should also count as a cycle in my opinion. (or if not, at least we have to document this).